faunadb-java/src/main/java/com/faunadb/client/types/Result.java

```diff
@@ -0,0 +1,140 @@
+package com.faunadb.client.types;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.base.Supplier;
+
+/**
+ * Represents the result of an operation. Usually a coercion operation.
+ *
+ * @see Codec
+ */
+public final class Result<T> {
```

Comment:
I've started with two classes and I end up merging them because the final implementation was much smaller this way... I can go back and split them again. I'm fine either way.